### PR TITLE
Add paste image

### LINKED ISSUE #10210: Paste an image in the editor
In GitHub it is possible to paste an image directly into the text input.
With the addition of Drag & Drop (#10145) I think it would be cool to also look at pasting images.

Here is the code the GitHub use <https://github.com/github/file-attachment-element> (This handles both Drag & Drop and Pasting but we'd only be interested in the latter).

I think it should be easy enough to extend @nickytonline's code with a new `utility` and a new `helper`.

### PR DESCRIPTION
**Pull request by Link2Twenty**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Add ability to paste images along side drag and drop

## Related Tickets & Documents
closes #10210

## QA Instructions, Screenshots, Recordings

forem__paste-image-directly

1. Copy an image into your clipboard.
2. Paste it into article body
3. See it appear

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [x] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?
N/A

## [optional] What gif best describes this PR or how it makes you feel?

Mad Scientist

### CODE REVIEW

**Comment by Link2Twenty** [comment]
@nickytonline I think this is ready now, sorry it took a while 😅

**Comment by MarcelloTheArcane** [comment]
DO IT DO IT DO IT! 🎉

**Review by ludwiczakpawel — APPROVED** [review]
I've merged the latest `master` (even though my commits history is a bit weird..?). It works perfect! Great job! I apologize it slipped out of our attention :C

**Review by katiedavis — APPROVED** [review]
Great!

File: app/javascript/utilities/pasteImage.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    element.addEventListener('paste', onPaste);
+
+    return () => {
+      element.removeEventListener('paste', onPaste);
+    };
+  }, [element, onPaste]);
+
+  return { setElement };
+}
```
Comment (nickytonline) [inline app/javascript/utilities/pasteImage.js:39]: Refactor (blocking): We're only returning a setter function. There is no need to return it in an object.

```suggestion
export function usePasteImage({ onPaste }) {
  const [element, setElement] = useState(null);

  useEffect(() => {
    if (!element) return;

    element.addEventListener('paste', onPaste);

    return () => {
      element.removeEventListener('paste', onPaste);
    };
  }, [element, onPaste]);

  return setElement;
}
```

File: app/javascript/article-form/components/EditorBody.jsx [hunk 2]
```diff
@@ -50,9 +52,17 @@ export const EditorBody = ({
     onDragExit,
   });
 
+  const setPasteElement = usePasteImage({
+    onPaste: handleImagePasted(
+      handleImageSuccess(textAreaRef),
+      handleImageFailure,
+    )
+  }).setElement;
```
Comment (nickytonline) [inline app/javascript/article-form/components/EditorBody.jsx:60]: Refactor (blocking): Ties into my refactor below.

```suggestion
  const setPasteElement = usePasteImage({
    onPaste: handleImagePasted(
      handleImageSuccess(textAreaRef),
      handleImageFailure,
    )
  });
```

**Review by nickytonline — CHANGES_REQUESTED** [review]
This looks great @Link2Twenty. One small request change. Looking forward to see this in action!

**Review by nickytonline — CHANGES_REQUESTED** [review]
@Link2Twenty I was testing this in Firefox and it throws an error. I believe we just need to check if the name property exists or not and if it doesn't just give it a default name like `"image"`.

image

**Comment by nickytonline** [comment]
@Link2Twenty, I've merged the latest from the main branch. Please grab latest of your branch. Thanks!

**Comment by Link2Twenty** [comment]
I've been having issues with gitpod which have prevented me from being able to work out the firefox issue.

**Comment by nickytonline** [comment]
One of the folks on the front-end team can take a peek at the Firefox issue if you want. It should be a pretty simple check to add an image name if there is none. cc: @forem/frontend

**Comment by aitchiss** [comment]
@Link2Twenty @nickytonline I've added a small change to work around the firefox issue - if the file detail array is empty we now add "alt text" as the alt, matching what we auto-populate through the manual image upload :)

**Review by nickytonline — APPROVED** [review]
Andrew, @aitchiss made the fix for Firefox, so this is good to go.

Thanks for the PR @Link2Twenty! 🚢

**Review by nickytonline — CHANGES_REQUESTED** [review]
Oops, looks like we need some tests here. @Link2Twenty, do you still need help with the testing?

**Comment by aitchiss** [comment]
@Link2Twenty @nickytonline I've added some tests for the hook itself and one of the helper functions. The other `handleImagePasted` helper function relies on the paste event - we could create a mock/fake event but I'm not sure we'd get much value from testing it that way (since we'd just be validating against our own assumptions of what that event should look like). Let me know if you disagree / can see another way though 🙂

**Review by nickytonline — APPROVED** [review]
Thanks so much for the PR @Link2Twenty and thanks @aitchiss forthe tests! 👏🏻🔥

**Pull request merged**